torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -149,36 +148,47 @@ void ProcessGroupAgent::enqueueSend(RpcWork work) {
           {
             (int64_t)pg_->getRank(),
             (int64_t)str.length(),
+            (int64_t)work.message_.type()
           }, {torch::kLong})
       };
 
       // ProcessGroup is not thread-safe when sending with the same tag, hence
       // the lock
-      std::unique_lock<std::mutex> lock(sendMutexes_.get()[work.rank_]);
-      pg_->send(preamble, work.rank_, work.rank_ /* channelTag */)->wait();
-      std::vector<torch::Tensor> payload =
-          {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
-      pg_->send(payload, work.rank_, work.rank_ /* channelTag */)->wait();
-      lock.unlock();
+      std::lock_guard<std::mutex> lock(sendMutexes_[work.to_]);
+      pg_->send(preamble, work.to_, work.to_ /* channelTag */)->wait();
+      if (work.message_.type() != MessageType::SHUTDOWN) {
+        std::vector<torch::Tensor> payload ={
```

Comment:
Why don't we build the payload outside the lock? In the common case, this would avoid extra operations under the lock. In the `shutdown` case, we can build an empty payload.